viewing_party/party.py

```diff
@@ -1,23 +1,212 @@
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+# calculates a user's average rating for movies from their watched list (let's see how their taste is)
+# parameters: user_data - a user's movie information including watched list
+# returns float average rating, 0.0 if empty watched list
+def get_watched_avg_rating(user_data):
+    # set the avg rating to 0 initially, if they have no movies in their watched list this is the default value returned
+    average_rating = 0.0
+    # make sure that their watched list is not empty because we will later be dividing by the length of it (we dont want a 0 denominator)
+    if user_data["watched"]:
+        # loop through each movie in the watched list and access its rating, add the rating to the average rating and divide by the length after iterating
+        for movie in user_data["watched"]:
+            average_rating += movie["rating"]
+        average_rating = average_rating / len(user_data["watched"])
+    return average_rating
+
+# finds the users most watched movie genre and returns it.
+# parameters: user_data - a user's movie information including watched list
+# returns string of most watched genre, None if empty watched list
+def get_most_watched_genre(user_data):
+    # create an empty dictionary that will hold the genres and the their occurences in the watched list
+    genre_dict = {}
+    # check that the watched list has movies
+    if user_data["watched"]:
+        # loop through and add each genre to the dictionary, if they already are in the dictionary then increment their dictionary value by 1
+        for movie in user_data["watched"]:
+            if movie["genre"] in genre_dict:
+                genre_dict[movie["genre"]] += 1
+            else:
+                genre_dict[movie["genre"]] = 1
+        # to find the genre with the highest count, initiate a variable for the highest watched genre and its count
+        # then loop through the genre dictionary and compare each genre's count to the stored value for count, if it is higher replace the count value with the genre's count value and the genre string with the respective genre
+        most_watched_genre = ""
+        count_for_genre = 0
+        for genre, quantity in genre_dict.items():
+            if quantity > count_for_genre:
+                count_for_genre = quantity
+                most_watched_genre = genre
+        return most_watched_genre
+
+    else:
+        return None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+# get a list of movies that you have watched but none of your friends have watched
+# parameters: users data
+# return list of movies objects
+def get_unique_watched(users_data):
+    friends_movies_set = []
+    users_movies_set = []

```

Comment:
Rather than first defining these as lists, adding everything, and then converting to sets, you could be a bit more efficient by starting out as sets and using the `add` method (rather than the `append` method for lists).

Reply:
This advice also applies to the `get_friends_unique_watched` function down below, but I won't repeat myself.